lib/rules/prefer-regex-literals.js

```diff
@@ -170,7 +311,83 @@ module.exports = {
                             context.report({ node, messageId: "unexpectedRedundantRegExp" });
                         }
                     } else if (hasOnlyStaticStringArguments(node)) {
-                        context.report({ node, messageId: "unexpectedRegExp" });
+                        let regexContent = getStringValue(node.arguments[0]);
+                        let noFix = false;
+                        let flags;
+
+                        if (node.arguments[1]) {
+                            flags = getStringValue(node.arguments[1]);
+                        }
+
+                        const regexppEcmaVersion = getRegexppEcmaVersion(context.parserOptions.ecmaVersion);
+                        const RegExpValidatorInstance = new RegExpValidator({ ecmaVersion: regexppEcmaVersion });
+
+                        try {
+                            RegExpValidatorInstance.validatePattern(regexContent, 0, regexContent.length, flags ? flags.includes("u") : false);
+                            if (flags) {
+                                RegExpValidatorInstance.validateFlags(flags);
+                            }
+                        } catch {
+                            noFix = true;
+                        }
+
+                        const tokenBefore = sourceCode.getTokenBefore(node);
+
+                        if (tokenBefore && !validPrecedingTokens.includes(tokenBefore.value)) {
+                            noFix = true;
+                        }
+
+                        if (!/^[-a-zA-Z0-9\\[\](){} \t\r\n\v\f!@#$%^&*+^_=/~`.><?,'"|:;]*$/u.test(regexContent)) {
+                            noFix = true;
+                        }
+
+                        if (regexContent && !noFix) {
+                            let charIncrease = 0;
+
+                            const ast = new RegExpParser({ ecmaVersion: regexppEcmaVersion }).parsePattern(regexContent, 0, regexContent.length, flags ? flags.includes("u") : false);
+
+                            visitRegExpAST(ast, {
+                                onCharacterEnter(characterNode) {
+                                    let changeCharIncrease = 0;
+                                    let stringCodePointValue = String.fromCodePoint(characterNode.value);
+                                    const escaped = resolveEscapes(characterNode.raw);
+
+                                    if (escaped) {
+                                        stringCodePointValue = escaped;
+                                        changeCharIncrease = characterNode.raw.length !== 1 ? 0 : 1;
+                                    }
+
+                                    regexContent =
+                                            regexContent.slice(0, characterNode.start + charIncrease) +
+                                            (escaped ? stringCodePointValue : characterNode.raw) +
+                                            regexContent.slice(characterNode.end + charIncrease);
```

Comment:
Can this be moved into `if (escaped) {}`? For those that don't need to be escaped, this just replaces `characterNode.raw` with itself, so it seems unnecessary?

Then, we could simplify `onCharacterEnter`:

```js
onCharacterEnter(characterNode) {
    const escaped = resolveEscapes(characterNode.raw);

    if (escaped) {
        regexContent =
            regexContent.slice(0, characterNode.start + charIncrease) +
            escaped +
            regexContent.slice(characterNode.end + charIncrease);
        charIncrease += escaped.length - characterNode.raw.length;
    }
}
```